File: agent/tools/crypto.py
"""Cryptocurrency tool using the free CoinGecko API.

Provides price lookup, market rankings, trending coins, coin details,
historical data, and global market statistics.
No API key required.
"""

import asyncio
import json
from typing import ClassVar, Type

import httpx
from cachetools import TTLCache
from langchain_core.tools import BaseTool
from pydantic import BaseModel, Field, PrivateAttr

from agent.logger import get_logger

logger = get_logger("agent.tools.crypto", component="crypto")

API_BASE = "https://api.coingecko.com/api/v3"
HEADERS = {"Accept": "application/json", "User-Agent": "OpenSentinel-Crypto/1.0"}


# =============================================================================
# Input Schema
# =============================================================================


class CryptoInput(BaseModel):
    action: str = Field(
        ...,
        description=(
            "Action to perform: "
            "'price' — get current price for one or more coins, "
            "'markets' — top coins by market cap, "
            "'search' — search for a coin by name/symbol, "
            "'trending' — currently trending coins, "
            "'detail' — detailed info for a single coin, "
            "'history' — historical price data, "
            "'global' — global market statistics."
        ),
    )
    ids: str = Field(
        default="",
        description=(
            "Comma-separated coin IDs for 'price' or single ID for 'detail'/'history'. "
            "Use CoinGecko IDs like 'bitcoin', 'ethereum', 'solana'. "
            "Use 'search' action to find the correct ID."
        ),
    )
    query: str = Field(default="", description="Search query for 'search' action.")
    limit: int = Field(default=20, ge=1, le=100, description="Number of results for 'markets' (default: 20).")
    days: int = Field(default=30, description="Number of days for 'history' (1, 7, 30, 90, 365).")
    currency: str = Field(default="usd", description="VS currency for prices (default: usd).")


# =============================================================================
# Tool Implementation
# =============================================================================


class CryptoTool(BaseTool):
    name: str = "crypto"
    description: str = (
        "Get cryptocurrency data: prices, market rankings, trending coins, "
        "coin details, historical prices, and global market stats. "
        "Uses the free CoinGecko API (no API key required).\n\n"
        "Examples:\n"
        "- Price check: action='price', ids='bitcoin,ethereum,solana'\n"
        "- Top coins: action='markets', limit=10\n"
        "- Search: action='search', query='cardano'\n"
        "- Trending: action='trending'\n"
        "- Coin details: action='detail', ids='bitcoin'\n"
        "- Price history: action='history', ids='bitcoin', days=30\n"
        "- Global stats: action='global'"
    )
    args_schema: Type[BaseModel] = CryptoInput
    handle_tool_error: bool = True

    _cache: TTLCache = PrivateAttr()
    _client: httpx.AsyncClient = PrivateAttr()

    MAX_CONTENT_CHARS: ClassVar[int] = 3000

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._cache = TTLCache(maxsize=200, ttl=300)  # 5 min cache
        self._client = httpx.AsyncClient(headers=HEADERS, timeout=15.0)

    async def _fetch(self, endpoint: str) -> dict:
        """Fetch JSON from CoinGecko API."""
        url = f"{API_BASE}{endpoint}"
        cache_key = url
        if cache_key in self._cache:
            return self._cache[cache_key]

        response = await self._client.get(url)
        if response.status_code == 429:
            raise RuntimeError("CoinGecko rate limit exceeded. Try again in a minute.")
        response.raise_for_status()
        data = response.json()
        self._cache[cache_key] = data
        return data

    async def _get_price(self, ids: str, currency: str) -> str:
        coin_ids = [i.strip().lower() for i in ids.split(",") if i.strip()]
        if not coin_ids:
            return json.dumps({"error": "Provide coin IDs (e.g., 'bitcoin,ethereum')"})

        data = await self._fetch(
            f"/simple/price?ids={','.join(coin_ids)}"
            f"&vs_currencies={currency}"
            f"&include_24hr_change=true&include_24hr_vol=true&include_market_cap=true"
        )

        prices = []
        for coin_id, info in data.items():
            prices.append({
                "id": coin_id,
                "price": info.get(currency),
                "change_24h": f"{info.get(f'{currency}_24h_change', 0):.2f}%",
                "volume_24h": info.get(f"{currency}_24h_vol"),
                "market_cap": info.get(f"{currency}_market_cap"),
            })

        return json.dumps({"currency": currency, "count": len(prices), "prices": prices}, indent=2)

    async def _get_markets(self, limit: int, currency: str) -> str:
        data = await self._fetch(
            f"/coins/markets?vs_currency={currency}&order=market_cap_desc"
            f"&per_page={limit}&page=1&sparkline=false"
        )

        coins = []
        for coin in data:
            coins.append({
                "rank": coin.get("market_cap_rank"),
                "id": coin.get("id"),
                "symbol": (coin.get("symbol") or "").upper(),
                "name": coin.get("name"),
                "price": coin.get("current_price"),
                "change_24h": f"{coin.get('price_change_percentage_24h', 0):.2f}%",
                "market_cap": coin.get("market_cap"),
                "volume_24h": coin.get("total_volume"),
            })

        return json.dumps({"currency": currency, "count": len(coins), "coins": coins}, indent=2)

    async def _search(self, query: str) -> str:
        if not query:
            return json.dumps({"error": "Provide a search query"})

        data = await self._fetch(f"/search?query={query}")
        coins = []
        for coin in (data.get("coins") or [])[:10]:
            coins.append({
                "id": coin.get("id"),
                "symbol": (coin.get("symbol") or "").upper(),
                "name": coin.get("name"),
                "rank": coin.get("market_cap_rank"),
            })

        return json.dumps({"query": query, "count": len(coins), "coins": coins}, indent=2)

    async def _get_trending(self) -> str:
        data = await self._fetch("/search/trending")
        coins = []
        for entry in data.get("coins", []):
            item = entry.get("item", {})
            item_data = item.get("data", {})
            coins.append({
                "id": item.get("id"),
                "symbol": (item.get("symbol") or "").upper(),
                "name": item.get("name"),
                "rank": item.get("market_cap_rank"),
                "price": item_data.get("price"),
                "change_24h": f"{(item_data.get('price_change_percentage_24h', {}).get('usd') or 0):.2f}%",
            })

        return json.dumps({"count": len(coins), "trending": coins}, indent=2)

    async def _get_detail(self, coin_id: str) -> str:
        if not coin_id:
            return json.dumps({"error": "Provide a coin ID (e.g., 'bitcoin')"})

        data = await self._fetch(
            f"/coins/{coin_id.strip().lower()}"
            f"?localization=false&tickers=false&community_data=false&developer_data=false"
        )

        md = data.get("market_data", {})
        result = {
            "id": data.get("id"),
            "symbol": (data.get("symbol") or "").upper(),
            "name": data.get("name"),
            "rank": data.get("market_cap_rank"),
            "description": (data.get("description", {}).get("en") or "")[:500],
            "categories": (data.get("categories") or [])[:5],
            "price": {
                "current": md.get("current_price", {}).get("usd"),
                "high_24h": md.get("high_24h", {}).get("usd"),
                "low_24h": md.get("low_24h", {}).get("usd"),
                "change_24h": f"{md.get('price_change_percentage_24h', 0):.2f}%",
                "change_7d": f"{md.get('price_change_percentage_7d', 0):.2f}%",
                "change_30d": f"{md.get('price_change_percentage_30d', 0):.2f}%",
            },
            "market": {
                "market_cap": md.get("market_cap", {}).get("usd"),
                "volume_24h": md.get("total_volume", {}).get("usd"),
                "circulating_supply": md.get("circulating_supply"),
                "total_supply": md.get("total_supply"),
                "max_supply": md.get("max_supply"),
            },
            "ath": {
                "price": md.get("ath", {}).get("usd"),
                "change": f"{md.get('ath_change_percentage', {}).get('usd', 0):.2f}%",
                "date": md.get("ath_date", {}).get("usd"),
            },
        }

        return json.dumps(result, indent=2)

    async def _get_history(self, coin_id: str, days: int) -> str:
        if not coin_id:
            return json.dumps({"error": "Provide a coin ID (e.g., 'bitcoin')"})

        data = await self._fetch(
            f"/coins/{coin_id.strip().lower()}/market_chart?vs_currency=usd&days={days}"
        )

        prices_raw = data.get("prices", [])
        prices = [
            {"date": p[0], "price": round(p[1], 2)}
            for p in prices_raw
        ]

        price_values = [p["price"] for p in prices]
        if price_values:
            first, last = price_values[0], price_values[-1]
            change = ((last - first) / first) * 100 if first else 0
            summary = {
                "start_price": first,
                "end_price": last,
                "change": f"{change:.2f}%",
                "high": max(price_values),
                "low": min(price_values),
            }
        else:
            summary = {}

        # Return last 30 data points to avoid token explosion
        return json.dumps({
            "id": coin_id,
            "days": days,
            "data_points": len(prices),
            "summary": summary,
            "prices": prices[-30:],
        }, indent=2)

    async def _get_global(self) -> str:
        data = await self._fetch("/global")
        d = data.get("data", {})

        return json.dumps({
            "active_cryptos": d.get("active_cryptocurrencies"),
            "markets": d.get("markets"),
            "total_market_cap": d.get("total_market_cap", {}).get("usd"),
            "total_volume_24h": d.get("total_volume", {}).get("usd"),
            "market_cap_change_24h": f"{d.get('market_cap_change_percentage_24h_usd', 0):.2f}%",
            "dominance": {
                "btc": f"{d.get('market_cap_percentage', {}).get('btc', 0):.2f}%",
                "eth": f"{d.get('market_cap_percentage', {}).get('eth', 0):.2f}%",
            },
        }, indent=2)

    # -------------------------------------------------------------------------
    # BaseTool interface
    # -------------------------------------------------------------------------

    async def _arun(
        self,
        action: str,
        ids: str = "",
        query: str = "",
        limit: int = 20,
        days: int = 30,
        currency: str = "usd",
    ) -> str:
        logger.info("crypto_action", action=action, ids=ids, query=query)
        try:
            match action.lower():
                case "price":
                    return await self._get_price(ids, currency)
                case "markets":
                    return await self._get_markets(limit, currency)
                case "search":
                    return await self._search(query)
                case "trending":
                    return await self._get_trending()
                case "detail":
                    return await self._get_detail(ids)
                case "history":
                    return await self._get_history(ids, days)
                case "global":
                    return await self._get_global()
                case _:
                    return json.dumps({
                        "error": f"Unknown action '{action}'. "
                        "Use: price, markets, search, trending, detail, history, global"
                    })
        except httpx.HTTPStatusError as e:
            logger.error("crypto_http_error", status=e.response.status_code)
            return json.dumps({"error": f"API error: HTTP {e.response.status_code}"})
        except Exception as e:
            logger.error("crypto_error", error=str(e))
            return json.dumps({"error": str(e)})

    def _run(
        self,
        action: str,
        ids: str = "",
        query: str = "",
        limit: int = 20,
        days: int = 30,
        currency: str = "usd",
    ) -> str:
        return asyncio.get_event_loop().run_until_complete(
            self._arun(action, ids, query, limit, days, currency)
        )

File: ts/crypto.ts
// MCP Crypto Server
// Cryptocurrency data using CoinGecko API (free, no API key required)
// https://www.coingecko.com/en/api

import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { z } from 'zod';

const API_BASE = 'https://api.coingecko.com/api/v3';

interface CryptoResult<T = unknown> {
    success: boolean;
    data?: T;
    error?: string;
}

async function fetchJson<T>(endpoint: string): Promise<CryptoResult<T>> {
    try {
        const response = await fetch(`${API_BASE}${endpoint}`, {
            headers: {
                Accept: 'application/json',
                'User-Agent': 'Flopsy-Crypto-MCP/1.0',
            },
        });

        if (!response.ok) {
            if (response.status === 429) {
                return { success: false, error: 'Rate limit exceeded. Try again in a minute.' };
            }
            return { success: false, error: `HTTP ${response.status}: ${response.statusText}` };
        }

        const data = (await response.json()) as T;
        return { success: true, data };
    } catch (err) {
        return { success: false, error: (err as Error).message };
    }
}

function formatResult<T>(result: CryptoResult<T>): {
    content: Array<{ type: 'text'; text: string }>;
} {
    if (!result.success) {
        return {
            content: [{ type: 'text', text: JSON.stringify({ error: result.error }) }],
        };
    }
    return {
        content: [{ type: 'text', text: JSON.stringify(result.data, null, 2) }],
    };
}

interface CoinPrice {
    [key: string]: {
        usd: number;
        usd_24h_change?: number;
        usd_24h_vol?: number;
        usd_market_cap?: number;
    };
}

async function getPrice(ids: string[], currency = 'usd'): Promise<CryptoResult> {
    const result = await fetchJson<CoinPrice>(
        `/simple/price?ids=${ids.join(',')}&vs_currencies=${currency}&include_24hr_change=true&include_24hr_vol=true&include_market_cap=true`,
    );

    if (!result.success) return result;

    const prices = Object.entries(result.data!).map(([id, data]) => ({
        id,
        price: data.usd,
        change24h: data.usd_24h_change?.toFixed(2) + '%',
        volume24h: data.usd_24h_vol,
        marketCap: data.usd_market_cap,
    }));

    return {
        success: true,
        data: {
            currency,
            count: prices.length,
            prices,
            timestamp: new Date().toISOString(),
        },
    };
}

interface MarketCoin {
    id: string;
    symbol: string;
    name: string;
    current_price: number;
    market_cap: number;
    market_cap_rank: number;
    price_change_percentage_24h: number;
    total_volume: number;
    high_24h: number;
    low_24h: number;
    circulating_supply: number;
    total_supply: number;
    ath: number;
    ath_change_percentage: number;
}

async function getMarkets(limit = 20, currency = 'usd'): Promise<CryptoResult> {
    const result = await fetchJson<MarketCoin[]>(
        `/coins/markets?vs_currency=${currency}&order=market_cap_desc&per_page=${limit}&page=1&sparkline=false`,
    );

    if (!result.success) return result;

    const coins = result.data!.map((coin) => ({
        rank: coin.market_cap_rank,
        id: coin.id,
        symbol: coin.symbol.toUpperCase(),
        name: coin.name,
        price: coin.current_price,
        change24h: coin.price_change_percentage_24h?.toFixed(2) + '%',
        marketCap: coin.market_cap,
        volume24h: coin.total_volume,
        high24h: coin.high_24h,
        low24h: coin.low_24h,
    }));

    return {
        success: true,
        data: {
            currency,
            count: coins.length,
            coins,
            timestamp: new Date().toISOString(),
        },
    };
}

interface SearchResult {
    coins: Array<{
        id: string;
        name: string;
        symbol: string;
        market_cap_rank: number;
        thumb: string;
    }>;
}

async function searchCoins(query: string): Promise<CryptoResult> {
    const result = await fetchJson<SearchResult>(`/search?query=${encodeURIComponent(query)}`);

    if (!result.success) return result;

    const coins = result.data!.coins.slice(0, 10).map((coin) => ({
        id: coin.id,
        symbol: coin.symbol.toUpperCase(),
        name: coin.name,
        rank: coin.market_cap_rank,
    }));

    return {
        success: true,
        data: {
            query,
            count: coins.length,
            coins,
        },
    };
}

interface TrendingResult {
    coins: Array<{
        item: {
            id: string;
            name: string;
            symbol: string;
            market_cap_rank: number;
            price_btc: number;
            data: {
                price: number;
                price_change_percentage_24h: { usd: number };
                market_cap: string;
            };
        };
    }>;
}

async function getTrending(): Promise<CryptoResult> {
    const result = await fetchJson<TrendingResult>('/search/trending');

    if (!result.success) return result;

    const coins = result.data!.coins.map((c) => ({
        id: c.item.id,
        symbol: c.item.symbol.toUpperCase(),
        name: c.item.name,
        rank: c.item.market_cap_rank,
        price: c.item.data?.price,
        change24h: c.item.data?.price_change_percentage_24h?.usd?.toFixed(2) + '%',
    }));

    return {
        success: true,
        data: {
            count: coins.length,
            trending: coins,
            timestamp: new Date().toISOString(),
        },
    };
}

interface CoinDetail {
    id: string;
    symbol: string;
    name: string;
    description: { en: string };
    links: { homepage: string[]; blockchain_site: string[] };
    market_cap_rank: number;
    market_data: {
        current_price: { usd: number };
        market_cap: { usd: number };
        total_volume: { usd: number };
        high_24h: { usd: number };
        low_24h: { usd: number };
        price_change_percentage_24h: number;
        price_change_percentage_7d: number;
        price_change_percentage_30d: number;
        circulating_supply: number;
        total_supply: number;
        max_supply: number;
        ath: { usd: number };
        ath_change_percentage: { usd: number };
        ath_date: { usd: string };
        atl: { usd: number };
        atl_date: { usd: string };
    };
    genesis_date: string;
    categories: string[];
}

async function getCoinDetail(id: string): Promise<CryptoResult> {
    const result = await fetchJson<CoinDetail>(
        `/coins/${id}?localization=false&tickers=false&community_data=false&developer_data=false`,
    );

    if (!result.success) return result;

    const coin = result.data!;
    const md = coin.market_data;

    return {
        success: true,
        data: {
            id: coin.id,
            symbol: coin.symbol.toUpperCase(),
            name: coin.name,
            rank: coin.market_cap_rank,
            description: coin.description.en?.slice(0, 500),
            website: coin.links.homepage[0],
            categories: coin.categories.slice(0, 5),
            genesisDate: coin.genesis_date,
            price: {
                current: md.current_price.usd,
                high24h: md.high_24h.usd,
                low24h: md.low_24h.usd,
                change24h: md.price_change_percentage_24h?.toFixed(2) + '%',
                change7d: md.price_change_percentage_7d?.toFixed(2) + '%',
                change30d: md.price_change_percentage_30d?.toFixed(2) + '%',
            },
            market: {
                marketCap: md.market_cap.usd,
                volume24h: md.total_volume.usd,
                circulatingSupply: md.circulating_supply,
                totalSupply: md.total_supply,
                maxSupply: md.max_supply,
            },
            ath: {
                price: md.ath.usd,
                changeFromAth: md.ath_change_percentage.usd?.toFixed(2) + '%',
                date: md.ath_date.usd,
            },
            atl: {
                price: md.atl.usd,
                date: md.atl_date.usd,
            },
            timestamp: new Date().toISOString(),
        },
    };
}

interface ChartData {
    prices: [number, number][];
}

async function getHistory(id: string, days: number): Promise<CryptoResult> {
    const result = await fetchJson<ChartData>(
        `/coins/${id}/market_chart?vs_currency=usd&days=${days}`,
    );

    if (!result.success) return result;

    const prices = result.data!.prices.map(([timestamp, price]) => ({
        date: new Date(timestamp).toISOString().split('T')[0],
        price: price,
    }));

    // Calculate stats
    const priceValues = prices.map((p) => p.price);
    const firstPrice = priceValues[0];
    const lastPrice = priceValues[priceValues.length - 1];
    const change = lastPrice && firstPrice ? ((lastPrice - firstPrice) / firstPrice) * 100 : 0;

    return {
        success: true,
        data: {
            id,
            days,
            dataPoints: prices.length,
            summary: {
                startPrice: firstPrice?.toFixed(2),
                endPrice: lastPrice?.toFixed(2),
                change: change.toFixed(2) + '%',
                high: Math.max(...priceValues).toFixed(2),
                low: Math.min(...priceValues).toFixed(2),
            },
            prices: prices.slice(-30), // Last 30 points to avoid too much data
        },
    };
}

interface GlobalData {
    data: {
        active_cryptocurrencies: number;
        markets: number;
        total_market_cap: { usd: number };
        total_volume: { usd: number };
        market_cap_percentage: { btc: number; eth: number };
        market_cap_change_percentage_24h_usd: number;
    };
}

async function getGlobalStats(): Promise<CryptoResult> {
    const result = await fetchJson<GlobalData>('/global');

    if (!result.success) return result;

    const d = result.data!.data;

    return {
        success: true,
        data: {
            activeCryptos: d.active_cryptocurrencies,
            markets: d.markets,
            totalMarketCap: d.total_market_cap.usd,
            totalVolume24h: d.total_volume.usd,
            marketCapChange24h: d.market_cap_change_percentage_24h_usd?.toFixed(2) + '%',
            dominance: {
                btc: d.market_cap_percentage.btc?.toFixed(2) + '%',
                eth: d.market_cap_percentage.eth?.toFixed(2) + '%',
            },
            timestamp: new Date().toISOString(),
        },
    };
}

export function createCryptoMcpServer() {
    const server = new McpServer({
        name: 'crypto-server',
        version: '1.0.0',
    });

    server.registerTool(
        'crypto_price',
        {
            title: 'Crypto Price',
            description:
                'Get current price for one or more cryptocurrencies. Use coin IDs like "bitcoin", "ethereum", "solana".',
            inputSchema: {
                ids: z
                    .array(z.string())
                    .describe("Coin IDs (e.g., ['bitcoin', 'ethereum', 'solana'])"),
            },
        },
        async ({ ids }) => {
            const result = await getPrice(ids.map((id) => id.toLowerCase()));
            return formatResult(result);
        },
    );

    server.registerTool(
        'crypto_markets',
        {
            title: 'Top Cryptocurrencies',
            description:
                'Get top cryptocurrencies by market cap with price, volume, and 24h change.',
            inputSchema: {
                limit: z.number().default(20).describe('Number of coins (default: 20, max: 100)'),
            },
        },
        async ({ limit }) => {
            const result = await getMarkets(Math.min(limit, 100));
            return formatResult(result);
        },
    );

    server.registerTool(
        'crypto_search',
        {
            title: 'Search Crypto',
            description:
                'Search for cryptocurrencies by name or symbol. Returns coin IDs for other tools.',
            inputSchema: {
                query: z.string().describe("Search term (e.g., 'bitcoin', 'eth', 'solana')"),
            },
        },
        async ({ query }) => {
            const result = await searchCoins(query);
            return formatResult(result);
        },
    );

    server.registerTool(
        'crypto_trending',
        {
            title: 'Trending Crypto',
            description: 'Get currently trending cryptocurrencies based on search popularity.',
            inputSchema: {},
        },
        async () => {
            const result = await getTrending();
            return formatResult(result);
        },
    );

    server.registerTool(
        'crypto_detail',
        {
            title: 'Crypto Details',
            description:
                'Get detailed info for a cryptocurrency: description, market data, ATH/ATL, supply.',
            inputSchema: {
                id: z
                    .string()
                    .describe(
                        "Coin ID (e.g., 'bitcoin', 'ethereum'). Use crypto_search to find IDs.",
                    ),
            },
        },
        async ({ id }) => {
            const result = await getCoinDetail(id.toLowerCase());
            return formatResult(result);
        },
    );

    server.registerTool(
        'crypto_history',
        {
            title: 'Crypto History',
            description: 'Get historical price data for a cryptocurrency.',
            inputSchema: {
                id: z.string().describe("Coin ID (e.g., 'bitcoin')"),
                days: z
                    .enum(['1', '7', '30', '90', '365'])
                    .default('30')
                    .describe('Number of days (default: 30)'),
            },
        },
        async ({ id, days }) => {
            const result = await getHistory(id.toLowerCase(), parseInt(days));
            return formatResult(result);
        },
    );

    server.registerTool(
        'crypto_global',
        {
            title: 'Global Crypto Stats',
            description:
                'Get global cryptocurrency market statistics: total market cap, volume, BTC dominance.',
            inputSchema: {},
        },
        async () => {
            const result = await getGlobalStats();
            return formatResult(result);
        },
    );

    return server;
}

async function main() {
    const server = createCryptoMcpServer();

    const transport = new StdioServerTransport();
    await server.connect(transport);
    console.error('[MCP] Crypto started');
}

// Run if executed directly
const isMain = import.meta.url === `file://${process.argv[1]}`;
if (isMain) {
    main().catch((error) => {
        console.error('[MCP Crypto Server] Fatal error:', error);
        process.exit(1);
    });
}
